src/components/TwoLevelSelector.js

```diff
@@ -0,0 +1,377 @@
+const React = require('react');
+const ReactDOM = require('react-dom');
+const Radium = require('radium');
+
+const Icon = require('./Icon');
+const StyleConstants = require('../constants/Style');
+
+const TwoLevelSelector = React.createClass({
+
+  propTypes: {
+    data: React.PropTypes.object,
+    onChange: React.PropTypes.func,
+    selectedItem: React.PropTypes.string,
+    title: React.PropTypes.string
+  },
+
+  getDefaultProps () {
+    return {
+      data: null,
+      onChange () {},
+      selectedItem: null,
+      title: 'Choose an item'
+    };
+  },
+
+  getInitialState () {
+    return {
+      childListTopOffset: 0,
+      highlightedItem: null,
+      searchKeyword: null,
+      searchList: [],
+      selectedItem: this.props.selectedItem || null,
+      showingOptions: false
+    };
+  },
+
+  _getChildItems (parentItem) {
+    return this.props.data[parentItem];
+  },
+
+  _getParentItems () {
+    const parentItems = [];
+    const items = this.props.data;
+
+    for (const key in items) {
+      if (items.hasOwnProperty(key)) {
+        parentItems.push(key);
+      }
+    }
+    return parentItems;
+  },
+
+  _flattenObjectToArray (obj) {
+    const tempArray = [];
+
+    for (const key in obj) {
+      if (obj.hasOwnProperty(key)) {
+        tempArray.push(key);
+      }
+
+      for (let i = 0; i < obj[key].length; i++) {
+        tempArray.push(obj[key][i]);
+      }
+    }
+    return tempArray;
+  },
+
+  _getCategoriesByKeyword (searchKeyword) {
+    const flatList = this._flattenObjectToArray(this.props.data);
+    const searchList = flatList.filter(option => {
+      return option.toLowerCase().match(searchKeyword.toLowerCase());
+    });
+
+    this.setState({
+      hoverItem: null,
+      searchList: searchKeyword ? searchList : [],
+      searchKeyword
+    });
+  },
+
+  _toggleShowingCategories () {
+    this.setState({
+      showingOptions: !this.state.showingOptions
+    });
+  },
+
+  _handleMouseOver (hoverItem, e) {
+    const marginTop = ReactDOM.findDOMNode(this.refs.anchorDiv).getBoundingClientRect().top;
+    const marginBottom = 25;
+    const li = ReactDOM.findDOMNode(e.target);
+    const liPosRelativeToViewport = li.getBoundingClientRect();
+    let topOffset = liPosRelativeToViewport.top - marginTop;
+    const childListCount = this._getChildItems(hoverItem).length;
+    const totalHeight = topOffset + (childListCount + 1) * li.offsetHeight;
+    const pageHeight = e.target.offsetParent.clientHeight;
+
+    if (totalHeight >= pageHeight) {
+      topOffset -= totalHeight - pageHeight - marginBottom;
+    }
+
+    this.setState({
+      hoverItem,
+      childListTopOffset: topOffset
+    });
+  },
+
+  _handleBlur () {
+    this.setState({
+      showingOptions: false
+    });
+  },
+
+  _handleClick (selectedItem) {
+    this.props.onChange(selectedItem);
```

Comment:
And make sure the selectedItem you are returning is the entire object and not just the label as a string.

```javascript
{ label: 'groceries', value: 1 }
```